### DISCUSSION TITLE
Alerting support for queries using transformations

### DISCUSSION THREAD

**Discussion by madan** [grafana/grafana#42881] [Alerting]
**What would you like to be added**:
Alerts for transformed metrics

**Why is this needed**:
We query extensively through stackdriver datasource, Grafana v7 feature of transformation helps a lot in gettig the right metric visualized, but without an alert being triggered based on a calculated value like the percentage of drops and so on, it doesn't add much value without the alerts

**Comment by eamonryan** [comment]
Seems relevant to #25681

**Comment by zeraphin21** [comment]
+1

**Comment by tobiangerstein** [comment]
+1

**Comment by torkelo** [comment]
Transformations are really a visualization tool, so you can reuse the same query result in many panels and filter, rename and order the result. 

Alerting happens on the backend so cannot use transformations.

**Comment by Gauravshah** [comment]
a lot of alerts are related to the calculations. It will be useful to have alerting. I thought the gel-app helps do this for transform

**Comment by PrestonTaylor** [comment]
> Transformations are really a visualization tool, so you can reuse the same query result in many panels and filter, rename and order the result.
> 
> Alerting happens on the backend so cannot use transformations.

Transformations are conceptually no different than the alert conditions feature, in fact beefing up the alert conditions calculation options would probably alleviate most people's needs here. For example I want to alert on a sum/avg/median of separate queries. That can be done by adding another query but that's unnecessary complexity and load on my data source. There's no technical reason why an alert which conveniently already has an evaluation interval couldn't evaluate from a transform. A transform will still return a value that can be greater than or less than X just like any query. There's also no reason why alerts couldn't work with templates, simply evaluate/enumerate the variables when you need to test the alert conditions. These two things alone would completely eliminate my need for using something like grafanalib and/or sad interns and would amplify the power of templating massively.

**Comment by jontro** [comment]
I would also like to add a use case for alerting on transformations. We want to alert on delta using different data sources. This is not possible currently since they need to be two separate queries.

**Comment by Gauravshah** [comment]
seems expressions are the way to go https://grafana.com/docs/grafana/latest/panels/expressions/

**Comment by armandgrillet** [comment]
Converting this to a discussion as this is a feature request and using expressions could be a workaround.

**Comment by joeblubaugh** [comment]
Hi all - we're discussing on the team what the effort required to support transformations as an alert rule expression would be.

**Comment by drhaz** [comment]
+1 for alerting on transformation. Our use case is to alert on a temperature-compensated pressure value, and that would require computation using two data sources (raw pressure reading + temperature reading from a different sensor).

**Reply by joeblubaugh** [reply]
How does using a transformation make this simpler than doing a computation with two data sources in the query editor?  In either case you do need to query both data sources, right?

**Comment by sholinaty** [comment]
Transforms are REQUIRED in the alerting tool, specifically for JSON-logs.

I pass in a JSON to the dashboard, which has (active: 3, total:6)
I want to graph those? I need to use the ExtractFields transform.
I want to alert if active < total... I need that ExtractFields transform.

**Comment by adamquan** [comment]
+1 American Airline needs this too.

**Comment by tmargono** [comment]
+1 Need this feature

**Comment by cheekp** [comment]
+1 My use case is to create an alert on an existing dashboard which uses the 'Mixed Data Source - Native Plugin' in combination with a Merge transform to check for differences between two data sets.

**Comment by Yi666** [comment]
+1 If alerting features can be supported on transformations, it will save much time for me to set up functional alerts based on the log data from somewhere else.

**Comment by umpa385** [comment]
adding to this as well transforms would be very helpful to have. expressions help, but not enough compared to transforms which allow us breaking out fields and renaming them with organize fields and extract fields.

**Comment by LucasTBC** [comment]
+1 for alerting based on transformations!!!!!!! Honestly, it would be really nice to be able to just create alert thresholds right in the Panel editing UI.

**Reply by LucasTBC** [reply]
Honestly, not having this might kill our Grafana use case outright....

**Comment by tstibbs** [comment]
One impact of not having this feature is that it increases the amount of config work required. I think anyone who's new to Grafana would assume that they could configure up a bunch of visualisations and then say "if this graph goes above this point, alert me". But it's far from this simple. 

What they actually have to do is configure up the visualisations, and then _separately_ create alerts. Sure, there's a button which creates an alert from a panel, but all this actually does is copy the viz query into the alert UI. The lack of transforms means that often the queries have to be written differently which makes it very difficult to keep them in sync with the queries behind the visualisations. Ideally we'd drive the alerts directly off the visualisations, but even if we can't do that, if we had transforms support for alerting then it would at least allow the queries behind the alerts to be exactly the same as the queries behind the visualisations, which would reduce the work to configure/maintain grafana.

**Comment by keidyz** [comment]
We have an Azure metrics query that returns a dynamic amount of time series which we flatten to one series by using a reduce transformation with the end goal to get the distinct count of time series.

Due to it being dynamic, we can't work with just expressions- at least with just the ones available right now.

Heavy +1 for introducing transformations to Alerts

**Comment by camathieu** [comment]
+1 from Criteo, we need mixed datasource query for Graphite -> PromQL migration and Tranforms are needed to cleanup the expression result (filter values).

**Comment by natellium** [comment]
Hey there! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

I will be closing this discussion, but if you’re still interested in this request being implemented please don’t hesitate to open a [feature request](https://github.com/grafana/grafana/issues/new?assignees=&labels=&projects=&template=1-feature_requests.md). Sorry for the inconvenience, and thanks for your contributions!

cc @grafana/alerting-squad

**Comment by bemorr** [comment]
Don't see this in an Github issue so created here : https://github.com/grafana/grafana/issues/81927 please feel free to point me at one created and close if I've missed it

**Comment by samjewell** [comment]
I've posted in the issue that @bemorr created too, but wanted to post an update here too:

We're actively working on this - specifically the SQL route that Leon mentioned above. You'll be able to add a SQL Expression, and then wrangle your query results to your heart's content. I'd recommend getting help from an LLM with your SQL, whether you're a beginner or intermediate SQL user: LLMs know SQL very well.

Feel free to follow https://github.com/grafana/grafana/pull/101820 where we are trying to refine the experience some more, before we start publicising this feature more widely.
